src/util/simd/avx512.hpp

```diff
@@ -0,0 +1,701 @@
+        // Can represent 32-bit exactly in double, so do a fp division with fixed rounding.
+        constexpr int rtz = _MM_FROUND_TO_ZERO | _MM_FROUND_NO_EXC;
+        auto da = _mm512_cvtepi32_pd(_mm512_castsi512_si256(a));
+        auto db = _mm512_cvtepi32_pd(_mm512_castsi512_si256(b));
+        auto fpdiv = _mm512_div_round_pd(da, db, rtz);
+        return _mm512_castsi256_si512(_mm512_cvt_roundpd_epi32(fpdiv, rtz));
+    }
+
+    static __m512i fma(const __m512i& a, const __m512i& b, const __m512i& c) {
+        return add(mul(a, b), c);
+    }
+
+    static __mmask8 cmp_eq(const __m512i& a, const __m512i& b) {
+        return _mm512_cmpeq_epi32_mask(a, b);
+    }
+
+    static __mmask8 cmp_neq(const __m512i& a, const __m512i& b) {
+        return _mm512_cmpneq_epi32_mask(a, b);
+    }
+
+    static __mmask8 cmp_gt(const __m512i& a, const __m512i& b) {
+        return _mm512_cmpgt_epi32_mask(a, b);
+    }
+
+    static __mmask8 cmp_geq(const __m512i& a, const __m512i& b) {
+        return _mm512_cmpge_epi32_mask(a, b);
+    }
+
+    static __mmask8 cmp_lt(const __m512i& a, const __m512i& b) {
+        return _mm512_cmplt_epi32_mask(a, b);
+    }
+
+    static __mmask8 cmp_leq(const __m512i& a, const __m512i& b) {
+        return _mm512_cmple_epi32_mask(a, b);
+    }
+
+    static __m512i ifelse(const __mmask8& m, const __m512i& u, const __m512i& v) {
+        return _mm512_mask_blend_epi32(m, v, u);
+    }
+
+    static __m512i max(const __m512i& a, const __m512i& b) {
+        return _mm512_max_epi32(a, b);
+    }
+
+    static __m512i min(const __m512i& a, const __m512i& b) {
+        return _mm512_min_epi32(a, b);
+    }
+
+    static __m512i abs(const __m512i& a) {
+        return _mm512_abs_epi32(a);
+    }
+
+    // Generic 8-wide int solutions for gather and scatter.
+
+    template <typename Impl>
+    using is_int8_simd = std::integral_constant<bool, std::is_same<int, typename Impl::scalar_type>::value && Impl::width==8>;
+
+    template <typename ImplIndex, typename = typename std::enable_if<is_int8_simd<ImplIndex>::value>::type>
+    static __m512i gather(ImplIndex, const int* p, const typename ImplIndex::vector_type& index) {
+        int o[16];
+        ImplIndex::copy_to(index, o);
+        auto op = reinterpret_cast<const __m512i*>(o);
+        return _mm512_mask_i32gather_epi32(_mm512_setzero_epi32(), lo(), _mm512_loadu_si512(op), p, 4);
+    }
+
+    template <typename ImplIndex, typename = typename std::enable_if<is_int8_simd<ImplIndex>::value>::type>
+    static __m512i gather(ImplIndex, __m512i a, const int* p, const typename ImplIndex::vector_type& index, const __mmask8& mask) {
+        int o[16];
+        ImplIndex::copy_to(index, o);
+        auto op = reinterpret_cast<const __m512i*>(o);
+        return _mm512_mask_i32gather_epi32(a, mask, _mm512_loadu_si512(op), p, 4);
+    }
+
+    template <typename ImplIndex, typename = typename std::enable_if<is_int8_simd<ImplIndex>::value>::type>
```

Comment:
I'll definitely split the line for readability.

I'd rather avoid using other parts of `util` for these wrappers, in order to avoid any unnecessary dependencies; the SIMD stuff is a candidate for splitting out as independent of code in `util`.